core/src/main/scala/akka/kafka/internal/DefaultProducerStage.scala

```diff
@@ -36,6 +36,13 @@ private[kafka] class DefaultProducerStage[K, V, P, IN <: Envelope[K, V, P], OUT
     new DefaultProducerStageLogic(this, inheritedAttributes)
 }
 
+private[kafka] object DefaultProducerStageLogic {
+  sealed trait ProducerAssignmentLifecycle
+  case object Unassigned extends ProducerAssignmentLifecycle
+  case object AsyncCreateRequestSent extends ProducerAssignmentLifecycle
+  case object Assigned extends ProducerAssignmentLifecycle
```

Comment:
Some API docs on these? Or somewhere else where the concept can be explained.

Reply:
I'll add some docs.  Basically it's used within the `TransactionalProducerStage` to handle messages differently depending on the state of the async producer assignment.  It's also useful to log state transitions for troubleshooting, so I included that too.  I incorporated it into your new trait `DeferredProducer`.